Support name mapping resolution for parquet

chenjunjiedada opened the pull request:
This fixes https://github.com/apache/incubator-iceberg/issues/715.

rdsr commented on parquet/src/main/java/org/apache/iceberg/parquet/ReadConf.java:
```diff
@@ -72,29 +74,31 @@
     if (filter != null) {
-      statsFilter = new ParquetMetricsRowGroupFilter(expectedSchema, filter, caseSensitive);
-      dictFilter = new ParquetDictionaryRowGroupFilter(expectedSchema, filter, caseSensitive);
+      NameMapping nameMapping = MappingUtil.create(ParquetSchemaUtil.convert(fileSchema));
```
I think the `nameMapping` should be taken as a parameter in `Parquet` as part of the builder and passed in from there

rdsr replied:
Also, if the `nameMapping` is not passed in and file schema does not have ids. We can infer the nameMapping, similar to what we've done here.  We had done this for Avro https://github.com/apache/incubator-iceberg/pull/580

chenjunjiedada replied:
It makes sense if we build name mapping from the expected schema, while here I need to build it from file schema so that we can use the metrics correctly.

rdsr commented on parquet/src/main/java/org/apache/iceberg/parquet/ReadConf.java:
```diff
@@ -72,29 +74,31 @@
-        ParquetSchemaUtil.pruneColumns(fileSchema, expectedSchema) :
-        ParquetSchemaUtil.pruneColumnsFallback(fileSchema, expectedSchema);
+            ParquetSchemaUtil.pruneColumns(fileSchema, expectedSchema) :
+            ParquetSchemaUtil.pruneColumnsByName(fileSchema, expectedSchema);
```
For Avro  we assigned ids to fileSchema based on the name mapping  and projected ids, which we derived from expected schema. [`Set<Integer> projectedIds = TypeUtil.getProjectedIds(expectedSchema)`] . This was pigged backed on column pruning code for Avro see - `AvroSchemaUtil.pruneColumns`

I'm unsure whether the same general approach can be applied here as well. @rdblue Thoughts?

chenjunjiedada replied:
Here is a reversed way. I build name mapping based on `fileSchema`, and then project ids for expected schema according to names. The reason is the metrics are built from parquet footer so we have to bind the expression to `fileSchema` instead of expected schema in `ParquetMetricsRowGroupFilter` and `ParquetDictionaryRowGroupFilter`.

chenjunjiedada commented:
@rdsr, I think again about your comments, that should be doable. I will submit an update later.

chenjunjiedada commented:
@rdsr, ready for another review.

chenjunjiedada commented on api/src/main/java/org/apache/iceberg/types/TypeUtil.java:
```diff
@@ -190,7 +191,7 @@ public static boolean isPromotionAllowed(Type from, Type.PrimitiveType to) {
 
   public static class SchemaVisitor<T> {
-    private final Deque<Integer> fieldIds = Lists.newLinkedList();
+    private final Deque<Integer> fieldIds = new ConcurrentLinkedDeque<>();
```
This is only used to pass the build, will fix in another pr.

chenjunjiedada commented:
@rdsr, any new comments? The new commit uses the way you suggest.

@rdblue, could you please help to take a look at your convenience as well?

rdsr commented:
@chenjunjiedada Been a little busy lately. I'll try to find time to look into it next week. Hope that's ok.

chenjunjiedada commented:
@rdsr no problem, take you time:)

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:
```diff
@@ -393,6 +395,11 @@ public ReadBuilder recordsPerBatch(int numRowsPerBatch) {
       return this;
     }
 
+    public ReadBuilder withNameMapping(NameMapping newNameMapping) {
```
This is called `nameMapping` in the Avro API. I think I prefer this name, but we should be consistent and either change that one or use the same name here.

chenjunjiedada replied:
Will change Avro side to use this one as well.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetDictionaryRowGroupFilter.java:
```diff
@@ -75,6 +86,16 @@ public ParquetDictionaryRowGroupFilter(Schema schema, Expression unbound, boolea
+      struct = schema.asStruct();
+    }
+
+    this.expr = Binder.bind(struct, Expressions.rewriteNot(expr), caseSensitive);
```
This class doesn't need to know about the name mapping. The mapping should be used to add IDs to the file schema so that most classes don't need to add specific support.

Look at how the other fallback happens using `ParquetSchemaUtil.addFallbackIds`. I think this should mimic that fallback in all cases.

chenjunjiedada replied:
The added unit test failed if we only use the ID fallback strategy. How about we keep both the original logic and the name mapping logic?

rdblue replied:
We need to keep both ID assignment strategies, but we don't need to change these filters. Just pass the file schema with assigned IDs into this like before.

chenjunjiedada replied:
I changed to pass `typeWithIds`.

rdblue replied:
Why is the name mapping still passed in here?

chenjunjiedada replied:
The evaluation build metadata mapping with `BlockMetadata` read from the footer which may contain columns that not in the file schema passed in, so I use name mapping here to filter out them.  The new commit changed to use a try/catch block.

rdblue replied:
The Parquet schema will match exactly because the IDs were added to the file schema. Adding IDs doesn't change the names, so `meta.getPath()` will be the same one used by `fileSchema.getType(...)`.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetMetricsRowGroupFilter.java:
```diff
@@ -64,8 +68,14 @@ public ParquetMetricsRowGroupFilter(Schema schema, Expression unbound) {
+
+  public ParquetMetricsRowGroupFilter withNameMapping(NameMapping newNameMapping) {
+    this.nameMapping = newNameMapping;
+    return this;
```
This file also should not be modified.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetReadSupport.java:
```diff
@@ -57,7 +58,7 @@ public ReadContext init(Configuration configuration, Map<String, String> keyValu
     MessageType projection = ParquetSchemaUtil.hasIds(fileSchema) ?
         ParquetSchemaUtil.pruneColumns(fileSchema, expectedSchema) :
-        ParquetSchemaUtil.pruneColumnsFallback(fileSchema, expectedSchema);
+        ParquetSchemaUtil.pruneColumnsByName(fileSchema, expectedSchema, MappingUtil.create(expectedSchema));
```
This should not replace the existing fallback strategy. If the mapping is present, it should be used. Otherwise, the existing position-based fallback strategy should be used.

chenjunjiedada replied:
Agreed.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ReadConf.java:
```diff
@@ -59,42 +61,46 @@
   private final boolean reuseContainers;
   @Nullable
   private final Integer batchSize;
+  @Nullable
```
No need for this annotation. Looks like we should have removed it in the other PR as well.

chenjunjiedada replied:
OK.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ReadConf.java:
```diff
@@ -59,42 +61,46 @@
-
     boolean hasIds = ParquetSchemaUtil.hasIds(fileSchema);
-    MessageType typeWithIds = hasIds ? fileSchema : ParquetSchemaUtil.addFallbackIds(fileSchema);
+    this.nameMapping = nameMapping == null ? MappingUtil.create(expectedSchema) : nameMapping;
```
This is not the right place to infer a name mapping. This class should apply the name mapping if it exists, and use a position-based fallback otherwise.

chenjunjiedada replied:
Agreed.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ReadConf.java:
```diff
@@ -59,42 +61,46 @@
-          statsFilter.shouldRead(typeWithIds, rowGroup) &&
-              dictFilter.shouldRead(typeWithIds, rowGroup, reader.getDictionaryReader(rowGroup)));
+          statsFilter.shouldRead(fileSchema, rowGroup) &&
+              dictFilter.shouldRead(fileSchema, rowGroup, reader.getDictionaryReader(rowGroup)));
```
What was the purpose of this change?

chenjunjiedada replied:
I reverted this back to `typeWithIds`

rdblue commented:
Thanks for working on this, @chenjunjiedada!

A few high-level things that I think need to be fixed:
* This should not replace the existing fallback that assigns IDs based on position. If the name mapping is configured, that should be used instead. Both are options. Feel free to do this by adding a name mapping to the existing `ParquetSchemaUtil` methods. We can do the fallback choice there.
* This also needs to update `ParquetSchemaUtil.hasIds` so that any ID causes it to return true, instead of any missing ID. This should be okay since files either have IDs or do not, so it is not a behavior change.
* As few classes should change as possible. Fallback ID assignment already works with the row group filters, so I don't think that we need to alter them. (Unless I'm missing some test case where it's failing!)

chenjunjiedada commented:
@rdblue,  I addressed some of the comments. The remaining needs a double check.

The ID fallback way cannot work when the iceberg table schema is different from the parquet file schema as the original issue mentioned. Also, the use case in the unit test (import partial data of existing parquet files as new iceberg table) would lead to the following exception which is related to metric row group filters. Therefore I removed the ID fallback way. I 'm not very sure that is the common use case, but if the name mapping way could handle more cases correctly, then it would be better to keep that way unless there is any case that it cannot handle. Does that make sense to you?

```
java.lang.ClassCastException: java.lang.Integer cannot be cast to java.lang.CharSequence
	at org.apache.iceberg.types.Comparators$CharSeqComparator.compare(Comparators.java:180)
	at org.apache.iceberg.types.Comparators$NullSafeChainedComparator.compare(Comparators.java:146)
	at org.apache.iceberg.parquet.ParquetMetricsRowGroupFilter$MetricsEvalVisitor.lt(ParquetMetricsRowGroupFilter.java:202)
	at org.apache.iceberg.parquet.ParquetMetricsRowGroupFilter$MetricsEvalVisitor.lt(ParquetMetricsRowGroupFilter.java:85)
	at org.apache.iceberg.expressions.ExpressionVisitors$BoundExpressionVisitor.predicate(ExpressionVisitors.java:122)
	at org.apache.iceberg.expressions.ExpressionVisitors.visitEvaluator(ExpressionVisitors.java:321)
	at org.apache.iceberg.expressions.ExpressionVisitors.visitEvaluator(ExpressionVisitors.java:336)
	at org.apache.iceberg.parquet.ParquetMetricsRowGroupFilter$MetricsEvalVisitor.eval(ParquetMetricsRowGroupFilter.java:108)
	at org.apache.iceberg.parquet.ParquetMetricsRowGroupFilter$MetricsEvalVisitor.access$000(ParquetMetricsRowGroupFilter.java:85)
	at org.apache.iceberg.parquet.ParquetMetricsRowGroupFilter.shouldRead(ParquetMetricsRowGroupFilter.java:79)
	at org.apache.iceberg.parquet.ReadConf.<init>(ReadConf.java:96)
	at org.apache.iceberg.parquet.ParquetReader.init(ParquetReader.java:64)
	at org.apache.iceberg.parquet.ParquetReader.iterator(ParquetReader.java:74)
	at org.apache.iceberg.spark.source.RowDataReader.open(RowDataReader.java:155)
	at org.apache.iceberg.spark.source.RowDataReader.open(RowDataReader.java:117)
	at org.apache.iceberg.spark.source.BaseDataReader.next(BaseDataReader.java:79)
	at org.apache.spark.sql.execution.datasources.v2.DataSourceRDD$$anon$1.hasNext(DataSourceRDD.scala:49)
	at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.sort_addToSorter_0$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$13$$anon$1.hasNext(WholeStageCodegenExec.scala:636)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:255)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:247)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/MessageTypeToType.java:
```diff
@@ -200,7 +200,7 @@ protected int nextId() {
   }
 
-  private int getId(org.apache.parquet.schema.Type type) {
+  protected int getId(org.apache.parquet.schema.Type type) {
```
Why was this changed?

chenjunjiedada replied:
This is changed so that we can override `getid` and use that to detect whether a schema has ID or not.

rdblue replied:
I think it might be better to make a new visitor for this, rather than reusing `MessageTypeToType`. That maintains independence between classes that don't have the same purpose. It was easy before to use `MessageTypeToType` without modification, but there's no need to do that if it is no longer easy.

chenjunjiedada replied:
OK, will add new one.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetSchemaUtil.java:
```diff
@@ -87,29 +82,21 @@ public static boolean hasIds(MessageType fileSchema) {
+        protected int getId(org.apache.parquet.schema.Type type) {
+          org.apache.parquet.schema.Type.ID id = type.getId();
+          if (id != null) {
+            throw new IllegalStateException("at least one ID exists");
```
It is fine if not everything has an ID. Columns without IDs will be ignored and should not cause a read to fail.

chenjunjiedada replied:
Did I misunderstand the comment?   I thought you meant file schema either has all columns with IDs or do not. Do you mean that a file schema can have some of the columns with IDs and some without ID?

> This also needs to update ParquetSchemaUtil.hasIds so that any ID causes it to return true, instead of any missing ID. This should be okay since files either have IDs or do not, so it is not a behavior change.

rdblue replied:
Yes, we expect schemas to either have all IDs or none, but we don't need to fail if that assumption is violated. If we detect any ID in the schema, then we should trust the file's IDs. If there are no IDs, then we should infer them using a name mapping or position-based fallback, whatever was configured.

rdblue commented:
> The remaining needs a double check.

I don't see replies to my comments. Can you reply so I know what you want me to double-check and why?

> The ID fallback way cannot work when the iceberg table schema is different from the parquet file schema as the original issue mentioned

Non-Iceberg table schemas are maintained either by position (like CSV) or by name (like JSON). The fallback strategy needs to match how schemas have been maintained for a table. We implemented position-based fallback because before Iceberg we maintained by position so we could rename fields and add fields, but not drop or reorder. Support for name mapping supports the name-based approach for maintaining a schema. Both approaches are valid.

You're right that some files are incompatible with position-based, but that just shows that you have to maintain a schema consistently. You can't use both position-based and name-based in a table.

chenjunjiedada commented:
@rdblue , thanks for the detail, I meant the comments about the ID fallback way.  I didn't take the CSV and JSON format into consideration. Now I understood the idea and will add ID fallback strategy back.

rdblue commented:
Great, thank you @chenjunjiedada!

chenjunjiedada commented:
@rdblue , I saw your comments while committing. You might want to see the new commit (with some rebasing commits.). I will address your lastest comments tomorrow.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/Avro.java:
```diff
@@ -235,7 +235,7 @@ public ReadBuilder rename(String fullName, String newName) {
     }
 
-    public ReadBuilder nameMapping(NameMapping newNameMapping) {
+    public ReadBuilder withNameMapping(NameMapping newNameMapping) {
```
@rdsr, just want to confirm that you're okay with this change? If it is going to cause you problems because you've already deployed it, we can go with the original.

rdblue commented:
@chenjunjiedada, it looks like lots of places have been changed to use the original file schema, rather than the file schema with fallback IDs added. I don't think that is correct.

Can you revert those changes, or help me understand why they are needed?

chenjunjiedada commented:
@rdblue , The reason for passing in name mapping in row group filter is because we build metadata mapping from footer which may contain incompatible columns with passing `typeWithIds`. I removed the name mapping through a try/catch block to solve that.

The original plan is to replace all ID-based `ParquetValueReader` to the name-based readers because the `NameBaseReadBuilder` builds a `ParquetValueReader` that doesn't need any schema ID. That is why I use the original file schema instead of `typeWithIds`. But there is no harm to pass a `MessageType` with IDs to build the name-based readers. so in the latest commit, I use `typeWithIds` for both ways.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetDictionaryRowGroupFilter.java:
```diff
@@ -108,11 +118,15 @@ private boolean eval(MessageType fileSchema, BlockMetaData rowGroup,
+            isFallback.put(id, ParquetUtil.hasNonDictionaryPages(meta));
+            mayContainNulls.put(id, mayContainNull(meta));
+          }
+        } catch (org.apache.parquet.io.InvalidRecordException e) {
```
Do you have a test case where this is thrown?

chenjunjiedada replied:
Yes. The `testImportWithIncompatibleSchema` in TestSparkTableUtil.java.

rdblue replied:
The problem here is that `ApplyNameMapping` is omitting fields that are in the file from the schema with field IDs. Instead of changing these filters, `ApplyNameMapping` should return all fields of the file schema, even if they are not assigned field IDs.

By doing that, we can assume that the file schema always matches the file metadata. That's important to avoid unexpected exceptions like this one: we're using the file's column metadata to look up a column in the schema and that should always work.

rdblue replied:
@chenjunjiedada, this needs to be fixed so that there are no changes to the filter implementations.

chenjunjiedada replied:
I tried to remove the `try...catch...` here and let `ApplyNameMapping` to return the schema with partial IDs assigned. That causes problems when we visit the parquet type.

Firstly, the `PruneColumns` visitor assumes the `typeWithIds` is fully assigned with IDs. Its `getId` has a precondition check. A file schema with partially assigned IDs cannot pass the condition check.

Secondly, the `HasId` logic also has the same assumption, which makes `SparkParquetReader` throws NPE when it calls `filedType.getId().intValue()`.

rdblue replied:
`HasId` should definitely succeed if there are fields without them. Its purpose is to check and we expect it to encounter schemas without IDs. Why is `HasId` being run on a schema that has had a name mapping applied?

If `PruneColumns` assumes all IDs are there, then I think we should fix that assumption.

chenjunjiedada replied:
`SparkParquetReader` build parquet reader according to `typeWithIds` passed from ReadConf.java. The `HasId` is used in ReadConf.java to assign IDs to file schema, and then it is used in `SparkParquetReader` to determine whether passing files schema has assigned IDs.  

Besides `PruneColumsn`, The `ReadBuilder` also assumes all IDs assigned.

chenjunjiedada replied:
@rdblue , do you think we need to change both `PruneColumns` and `SparkParquetReader#ReadBuilder` to satisfy the assumption? It may change a lot.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetMetricsRowGroupFilter.java:
```diff
@@ -76,6 +81,9 @@ public ParquetMetricsRowGroupFilter(Schema schema, Expression unbound, boolean c
    */
   public boolean shouldRead(MessageType fileSchema, BlockMetaData rowGroup) {
+    StructType struct = schema.asStruct();
+    this.expr = Binder.bind(struct, Expressions.rewriteNot(expr), caseSensitive);
```
Please revert unnecessary changes and move expression binding back into the constructor.

chenjunjiedada replied:
Sure.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetReadSupport.java:
```diff
@@ -55,9 +58,16 @@ public ReadContext init(Configuration configuration, Map<String, String> keyValu
+      if (nameMapping != null) {
+        projection = ParquetSchemaUtil.pruneColumnsByName(fileSchema, expectedSchema, nameMapping);
+      } else {
+        projection = ParquetSchemaUtil.pruneColumnsFallback(fileSchema, expectedSchema);
```
Let's simplify this logic by adding `nameMapping` to `pruneColumnsFallback`. Then inside that method, we can decide the fallback strategy based on whether nameMapping is null.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ReadConf.java:
```diff
@@ -66,19 +63,27 @@
-
     boolean hasIds = ParquetSchemaUtil.hasIds(fileSchema);
-    MessageType typeWithIds = hasIds ? fileSchema : ParquetSchemaUtil.addFallbackIds(fileSchema);
+    MessageType typeWithIds;
```
Like above, let's pass `nameMapping` to `addFallbackIds` and have that choose whether to assign using the mapping or by position.

chenjunjiedada replied:
OK.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetSchemaUtil.java:
```diff
@@ -112,4 +127,56 @@ public static MessageType addFallbackIds(MessageType fileSchema) {
   }
+
+  public static MessageType addFallbackIds(MessageType fileSchema, NameMapping nameMapping) {
+    MessageTypeBuilder builder = org.apache.parquet.schema.Types.buildMessage();
```
This is should recursively assign IDs using a visitor, not just assign to the top-level.

rdblue replied:
I think we need a test suite for this assignment, too.

chenjunjiedada replied:
Hmm. My bad. Will update in the next commit.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetSchemaUtil.java:
```diff
@@ -112,4 +127,56 @@ public static MessageType addFallbackIds(MessageType fileSchema) {
+    return builder.named(fileSchema.getName());
+  }
+
+  public static class HasIds extends ParquetTypeVisitor<Boolean> {
```
This looks good.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetReadSupport.java:
```diff
@@ -55,9 +58,16 @@ public ReadContext init(Configuration configuration, Map<String, String> keyValu
+      projection = ParquetSchemaUtil.pruneColumns(fileSchema, expectedSchema);
+    } else {
+      if (nameMapping != null) {
+        projection = ParquetSchemaUtil.pruneColumnsByName(fileSchema, expectedSchema, nameMapping);
```
When a name mapping is used, column pruning should happen using the normal `pruneColumns`. The reason why the other fallback path uses `pruneColumnsFallback` is because ordinal IDs can only be assigned to the top-level columns and so projection must only use top-level column IDs. That's not a limitation of name mapping, so we can use normal pruning once the IDs are filled in.

chenjunjiedada replied:
I got your point now. Will remove `pruneColumnsByName`.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetSchemaUtil.java:
```diff
@@ -82,23 +86,34 @@ public static MessageType pruneColumnsFallback(MessageType fileSchema, Schema ex
+   * @param expectedSchema expected schema
+   * @return a parquet schema pruned using the expected schema
+   */
+  public static MessageType pruneColumnsByName(MessageType fileSchema, Schema expectedSchema, NameMapping nameMapping) {
```
We don't need this method. Instead, let's add the IDs and then use normal schema pruning.

chenjunjiedada replied:
Agreed.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/data/SparkParquetReaders.java:
```diff
@@ -111,6 +124,49 @@ private SparkParquetReaders() {
     }
   }
 
+  private static class NameBasedReadBuilder extends ReadBuilder {
```
I don't think a name-based reader is necessary. Instead, the file schema passed to the builder should have IDs assigned.

chenjunjiedada replied:
I 'm OK to remove this to make the parquet reader consistent.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java:
```diff
@@ -171,10 +173,12 @@
+        .withNameMapping(nameMapping)
         .split(task.start(), task.length())
-        .createReaderFunc(fileSchema -> SparkParquetReaders.buildReader(readSchema, fileSchema))
+        .createReaderFunc(fileSchema -> SparkParquetReaders.buildReader(readSchema, fileSchema, nameMapping))
```
The `fileSchema` passed here should by the type with IDs filled in.

chenjunjiedada replied:
Agreed, I think we don't need the third parameter `namemapping` as well.

chenjunjiedada commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetTypeVisitor.java:
```diff
@@ -34,9 +34,13 @@
+    return visit(type, visitor, false);
+  }
+
+  public static <T> T visit(Type type, ParquetTypeVisitor<T> visitor, Boolean ignoreRepeated) {
```
I add an extra parameter here because the field names built from the original `ParquetTypeVisitor` are different in name mapping. For example, if we have a field consists of a list of the map type like below:
```
optional(1, "list_of_maps",Types.ListType.ofOptional(2, Types.MapType.ofOptional(3, 4,
    Types.StringType.get(),Types.StringType.get());
```
the field names of the map ke and value are `list_of_map.list.element.map.key` and `list_of_map.list.element.map.value`. While the names in name mapping are `list_of_map.element.key` and `list_of_map.element.value`.

Since the `NameMapping` is used by Avro already, so I update `ParquetTypeVisitor` a bit to handle this.

rdblue replied:
Thanks for the explanation, this will be useful when I take a closer look.

rdblue replied:
I think we should use an alternative approach. I opened #1001 for this.

The problem with this is that it adds a parameter to customize behavior, when that should be the responsibility of the visitor. This is also a bit confusing because it isn't clear why some callers would want to skip repeated fields. It is better to delegate this to the visitor.

The PR I opened allows you to do that. It adds before and after callbacks for the repeated fields (element and key_value), as well as for the element, key, and value types. By default, those call `beforeField` and `afterField`, which are used to maintain the field names list. With that change, you can simply override the repeated element and repeated key value calls to do nothing:

```java

  @Override
  public void beforeRepeatedElement(Type element) {
    // do not add the repeated element's name
  }

  @Override
  public void afterRepeatedElement(Type element) {
    // do not remove the repeated element's name
  }

  @Override
  public void beforeRepeatedKeyValue(Type keyValue) {
    // do not add the repeated element's name
  }

  @Override
  public void afterRepeatedKeyValue(Type keyValue) {
    // do not remove the repeated element's name
  }
```

That passes all of the tests.

chenjunjiedada replied:
Great! I will update after the PR merged.

chenjunjiedada commented:
@rdblue, Would you please take another look? It is ready.

chenjunjiedada commented:
@rdblue, Could you please take another look at this? This is not related to v2 things, can we add this to 0.8.0 release?

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetSchemaUtil.java:
```diff
@@ -83,33 +92,144 @@ public static MessageType pruneColumnsFallback(MessageType fileSchema, Schema ex
+
+    @Override
+    public Boolean list(GroupType array, Boolean hasId) {
+      if (hasId) {
```
Minor: this could be simplified to `return hasId || array.getId() != null`.

chenjunjiedada replied:
Done.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetSchemaUtil.java:
```diff
@@ -83,33 +92,144 @@ public static MessageType pruneColumnsFallback(MessageType fileSchema, Schema ex
-      return true;
+    @Override
+    public Boolean map(GroupType map, Boolean keyHasId, Boolean valueHasId) {
+      if (keyHasId || valueHasId) {
```
This could also be simplified to `return keyHasId || valueHasId || map.getId() != null`

chenjunjiedada replied:
Done.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetSchemaUtil.java:
```diff
@@ -83,33 +92,144 @@ public static MessageType pruneColumnsFallback(MessageType fileSchema, Schema ex
 
-  public static MessageType addFallbackIds(MessageType fileSchema) {
-    MessageTypeBuilder builder = org.apache.parquet.schema.Types.buildMessage();
+  public static class AssignIdsByNameMapping extends ParquetTypeVisitor<Type> {
```
This overwrites any existing IDs, so let's call it `applyNameMapping` instead.

Also, this is quite a large class. Can you move it into its own file? It should be package-private.

chenjunjiedada replied:
Done.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetSchemaUtil.java:
```diff
@@ -83,33 +92,144 @@ public static MessageType pruneColumnsFallback(MessageType fileSchema, Schema ex
     }
 
-    return builder.named(fileSchema.getName());
+    private String[] currentPath() {
```
Please rebase on master. #950 moved these methods into the `ParquetTypeVisitor` so you don't have to copy them.

chenjunjiedada replied:
Done.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetSchemaUtil.java:
```diff
@@ -83,33 +92,144 @@ public static MessageType pruneColumnsFallback(MessageType fileSchema, Schema ex
+    }
+
+    @Override
+    public Type message(MessageType message, List<Type> fields) {
```
Overall, the implementations in this class look correct to me.

chenjunjiedada replied:
Thanks.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ReadConf.java:
```diff
@@ -66,19 +63,21 @@
     this.options = options;
     this.reader = newReader(file, options);
     MessageType fileSchema = reader.getFileMetaData().getSchema();
-
```
Let's leave this line to avoid git conflicts.

chenjunjiedada replied:
ok.

rdblue commented on parquet/src/test/java/org/apache/iceberg/parquet/TestParquetSchemaUtil.java:
```diff
@@ -0,0 +1,95 @@
+    MessageType typeWithIdsFromNameMapping = ParquetSchemaUtil.addFallbackIds(messageType, nameMapping);
+    Schema newSchema = ParquetSchemaUtil.convert(typeWithIdsFromNameMapping);
+
+    Assert.assertEquals(schema.toString(), newSchema.toString());
```
This should use `asStruct()` instead of `toString`.

chenjunjiedada replied:
Done.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ReadConf.java:
```diff
@@ -66,19 +63,21 @@
-    MessageType typeWithIds = hasIds ? fileSchema : ParquetSchemaUtil.addFallbackIds(fileSchema);
+    MessageType typeWithIds = hasIds ? fileSchema : ParquetSchemaUtil.addFallbackIds(fileSchema, nameMapping);
+
+    if (nameMapping == null && !hasIds) {
```
This logic assumes what `addFallbackIds` is going to do. I think I'd rather structure this a bit differently:

```
MessageType typeWithIds;
if (ParquetSchemaUtil.hasIds(fileSchema)) {
  typeWithIds = fileSchema;
  this.projection = ParquetSchemaUtil.pruneColumns(fileSchema, expectedSchema);
} else if (nameMapping != null) {
  typeWithIds = ParquetSchemaUtil.applyNameMapping(fileSchema, nameMapping);
  this.projection = ParquetSchemaUtil.pruneColumns(typeWithIds, expectedSchema);
} else {
  typeWithIds = ParquetSchemaUtil.addFallbackIds(fileSchema);
  this.projection = ParquetSchemaUtil.pruneColumnsFallback(fileSchema, expectedSchema);
}
```

chenjunjiedada replied:
Done.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java:
```diff
@@ -192,8 +194,10 @@
       FileScanTask task,
       Schema readSchema,
       Map<Integer, ?> idToConstant) {
+    NameMapping nameMapping = MappingUtil.create(readSchema);
```
This shouldn't create a name mapping. It should use one from the table, if it exists.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetReadSupport.java:
```diff
@@ -55,9 +58,16 @@ public ReadContext init(Configuration configuration, Map<String, String> keyValu
-        ParquetSchemaUtil.pruneColumns(fileSchema, expectedSchema) :
-        ParquetSchemaUtil.pruneColumnsFallback(fileSchema, expectedSchema);
+    MessageType projection;
+
```
No need for a newline here.

chenjunjiedada replied:
Done

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/AssignIdsByNameMapping.java:
```diff
@@ -0,0 +1,102 @@
+import org.apache.parquet.schema.Type;
+import org.apache.parquet.schema.Types;
+
+class AssignIdsByNameMapping extends ParquetTypeVisitor<Type> {
```
Let's rename this to mach the method, `ApplyNameMapping`.

chenjunjiedada replied:
OK.

rdblue commented on spark/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java:
```diff
@@ -199,4 +204,41 @@ public void testImportAsHiveTable() throws Exception {
+
+    // The field is different so that it will project with name mapping
+    Schema filteredSchema = new Schema(
+            optional(1, "data", Types.StringType.get())
```
Nit: Indentation is incorrect here.

chenjunjiedada replied:
Done

rdblue commented on spark/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java:
```diff
@@ -199,4 +204,41 @@ public void testImportAsHiveTable() throws Exception {
+
+    TableIdentifier source = new TableIdentifier("original_table");
+    Table table = catalog.createTable(
+        org.apache.iceberg.catalog.TableIdentifier.of(DB_NAME, "target_table"),
```
Why isn't `TableIdentifier` imported?

chenjunjiedada replied:
`TableIdentifier` from spark is imported so that it uses the full qualified name to avoid conflict. We use it in other places in this file as well.

rdblue commented on spark/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java:
```diff
@@ -199,4 +204,41 @@ public void testImportAsHiveTable() throws Exception {
+    spark.table(qualifiedTableName).write().mode("overwrite").format("parquet")
+        .saveAsTable("original_table");
+
+    // The field is different so that it will project with name mapping
```
Name mapping should only be used if it is set on the table.

chenjunjiedada replied:
Done.

rdblue commented on spark/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java:
```diff
@@ -199,4 +204,41 @@ public void testImportAsHiveTable() throws Exception {
   }
+
+  @Test
+  public void testImportWithIncompatibleSchema() throws Exception {
```
This should be named `testImportWithNameMapping` instead. The schema isn't incompatible, it just uses a different strategy (name mapping, instead of position mapping) to assign IDs for fields.

chenjunjiedada replied:
Done.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/AssignIdsByNameMapping.java:
```diff
@@ -0,0 +1,102 @@
+  public Type list(GroupType list, Type elementType) {
+    MappedField field = nameMapping.find(currentPath());
+    if (field == null) {
+      return null;
```
This should return the list with IDs assigned to sub-types.

chenjunjiedada replied:
It was intended to return null since the name mapping is created from the table's schema. I changed to return list with sub-types now since the assumption is removed.

rdblue commented on spark/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java:
```diff
@@ -199,4 +204,41 @@ public void testImportAsHiveTable() throws Exception {
+        filteredSchema,
+        SparkSchemaUtil.specForTable(spark, "original_table"));
+    File stagingDir = temp.newFolder("staging-dir");
+    SparkTableUtil.importSparkTable(spark, source, table, stagingDir.toString());
```
When a Parquet file is imported, its schema is converted to Iceberg using `MessageTypeToType`, which assigns IDs to fields using the position-based strategy. Stats from those files are generated and stored in Iceberg metadata using the position-based IDs, not name-based. That means for this imported table, the stats in Iceberg metadata are incorrect.

To fix this, you also need to support name mapping in the conversion from Parquet to an Iceberg schema, and make sure that is called here. Alternatively, you could detect that the table has a name mapping and skip metrics when importing data files.

chenjunjiedada replied:
Where should the name mapping come from? I thought the name mapping should be created via the table's schema because the typical case is we create a table, import the parquet files from an existing spark table. In case of that, we only care about the newly created fields and their stats, that's why the visitor returns null when we cannot find a field in name mapping regardless of its sub-types  (it is impossible to define a schema with list/map/struct without a name and has the subfields).

If the name mapping comes from external, for example, from the converted schema of parquet file schema, then the logic would be different as you pointed out. We might need 1) store name mapping for Table, 2) a Table API to get name mapping, 3) change `MessageTypeToType` to visit schema with name mapping. That seems to bring in too many complexities.

Back to the issue itself, what we need is creating a name mapping when `hasIds` is false. So we can add another property to Parquet ReadBuilder, say `enableNameMapping`, then we could build a default name mapping from `expectedSchema` even though we don't pass a name mapping. The logic in the ReadConf.java would be:

```
NameMapping actualNameMapping;
if (enableNameMapping) {
  if(nameMapping != null) {
    actualNameMapping = nameMapping;
  } else {
    actualNameMapping = MappingUtil.create(expectedSchema);
  }
```

What do you think?

rdblue replied:
Name mapping should be part of table configuration. If one is present, it should be used. Otherwise no name mapping should be used.

chenjunjiedada replied:
Parsing name mapping make sense to me, that could be used for avro and orc as well.  While this would need to parse the string from table property to NameMapping, how about we make this as a follow-up and use a default name mapping created from table schema as I describe above?

rdblue replied:
No, we should not commit anything that creates a mapping and uses it automatically.

chenjunjiedada replied:
OK, I updated this.

chenjunjiedada commented:
@rdblue , I rebased this, could you please help to take another look? Thanks!

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:
```diff
@@ -419,17 +433,18 @@ public ReadBuilder recordsPerBatch(int numRowsPerBatch) {
       ParquetReadBuilder<D> builder = new ParquetReadBuilder<>(ParquetIO.file(file));
 
-      builder.project(schema);
+      builder.project(schema).withNameMapping(MappingUtil.create(schema));
```
Parquet should never automatically add a mapping. When reading, a name mapping should be parsed and added if one is present in table properties.

chenjunjiedada replied:
Agreed.

chenjunjiedada commented:
@rdblue , Since `PruneColumns` and `SparkParquetReader#ReaderBuilder` needs file schema with full assigned IDs, I revert`ApplyNameMapping` back to ignore the fields that don't have IDs to keep logic correct.

So we should either change `PruneColumns` and `SparkParquetReader#ReaderBuilder` visitor to accept a file schema with partial IDs assigned or use the `try...catch...` clause in the row group filter. What do you think?

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:
```diff
@@ -393,6 +396,16 @@ public ReadBuilder recordsPerBatch(int numRowsPerBatch) {
+      return this;
+    }
+
+    public ReadBuilder applyNameMapping() {
```
What is the purpose of this? Why not always apply a mapping if it was supplied?

I think it is confusing to only apply a mapping if this is called, even when the mapping is set.

chenjunjiedada replied:
The idea is if there is no name mapping supplied, we use a default name mapping build from table schema.

I 'm still a bit confused that why we would allow users to pass in a customized name mapping. IMHO, there are only two name mappings should be legal, either we create it from parquet file or we create it from the table schema. Is there any scenario that user would pass in other mappings?

chenjunjiedada replied:
I removed this and use name mapping from table property now.

chenjunjiedada commented:
@rdblue , I addressed the latest two comments about using the passing name mapping and allow parquet type without IDs. Would you please take another look? Thanks for your patience!

chenjunjiedada commented:
@rdblue , This is ready for another review. Thanks

chenjunjiedada commented:
Merged again. @rdblue, would you please take another look?

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetMetricsRowGroupFilter.java:
```diff
@@ -47,6 +47,7 @@
 import org.apache.parquet.schema.PrimitiveType;
 
 public class ParquetMetricsRowGroupFilter {
+
```
Looks like this is a non-functional change to an unrelated file. Could you revert it? Same with `ParquetDictionaryRowGroupFilter`.

chenjunjiedada replied:
Strange, I remember I fixed things like this, maybe the fix was reverted by merging.  Anyway, let me fix this.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/VectorizedParquetReader.java:
```diff
@@ -48,11 +49,12 @@
-      Function<MessageType, VectorizedReader<?>> readerFunc,
-      Expression filter, boolean reuseContainers, boolean caseSensitive, int maxRecordsPerBatch) {
+      InputFile input, Schema expectedSchema, ParquetReadOptions options, Function<MessageType,
+      VectorizedReader<?>> readerFunc, NameMapping nameMapping, Expression filter, boolean reuseContainers,
```
Please don't split types across lines. I think moving `Function<MessageType` to the previous line should be reverted.

rdblue commented on parquet/src/test/java/org/apache/iceberg/parquet/TestParquetSchemaUtil.java:
```diff
@@ -0,0 +1,95 @@
+        .asStructType().fields());
+    NameMapping nameMapping = MappingUtil.create(schema);
+    MessageType messageType = ParquetSchemaUtil.convert(schema, "complex_schema");
+    MessageType typeWithIdsFromNameMapping = ParquetSchemaUtil.applyNameMapping(messageType, nameMapping);
```
There's a problem here: `applyNameMapping` could be a noop and this would still work because the conversion to Parquet and back preserves IDs. For Avro, there is a test utility to remove IDs from the schema so that we can test adding them back with the name mapping. I think it would make sense to take the Avro name mapping tests and adapt them for Parquet as well. That can be done in a follow-up, but this test does need to be fixed before committing.

chenjunjiedada replied:
Make sense to me, let me file an issue to record this.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/PruneColumns.java:
```diff
@@ -45,13 +44,16 @@ public Type message(MessageType message, List<Type> fields) {
-        hasChange = true;
+      Integer fieldId = getId(originalField);
+      if (fieldId != null) {
+        if (selectedIds.contains(fieldId)) {
```
Why is this not `if (fieldId != null && selectedIds.contains(fieldId))`?

The else case is used when a sub-field is projected by ID. So the question is whether a sub-field can be projected if its parents aren't mapped. I think we should allow it because it would be confusing to have a value mapped, but still get nulls because a parent is not.

chenjunjiedada replied:
Make sense to me. Updated.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/PruneColumns.java:
```diff
@@ -71,11 +73,14 @@ public Type struct(GroupType struct, List<Type> fields) {
-        filteredFields.add(originalField);
-        hasChange = true;
+      Integer fieldId = getId(originalField);
+      if (fieldId != null) {
```
Same logic as above applies here.

rdblue replied:
And in the rest of the updates for this file.

chenjunjiedada replied:
Done.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/data/SparkParquetReaders.java:
```diff
@@ -142,9 +142,11 @@ private SparkParquetReaders() {
+          int id = fieldType.getId().intValue();
+          readersById.put(id, ParquetValueReaders.option(fieldType, fieldD, fieldReaders.get(i)));
+          typesById.put(id, fieldType);
+        }
```
Looks correct to me.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java:
```diff
@@ -65,8 +68,10 @@
+          .reuseContainers();
+
+      iter = nameMapping != null ?
+          builder.withNameMapping(NameMappingParser.fromJson(nameMapping)).build() : builder.build();
```
I'd prefer to keep the name mapping and building separate. There's no need to mix these together.

```java
if (nameMapping != null) {
  builder.withNameMapping(NameMappingParser.fromJson(nameMapping));
}

iter = builder.build();
```

rdblue commented on spark/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java:
```diff
@@ -200,4 +209,48 @@ public void testImportAsHiveTable() throws Exception {
   }
+
+  @Test
+  public void testImportWithNameMapping() throws Exception {
```
Can you add a test for name mapping with the vectorized read path? Just set the property on the table and it should take that path.

chenjunjiedada replied:
Done, also updated the vectorized reader builder. Thanks for catching this

rdblue commented on spark/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java:
```diff
@@ -200,4 +209,48 @@ public void testImportAsHiveTable() throws Exception {
+    );
+
+    NameMapping nameMapping = MappingUtil.create(filteredSchema);
+
```
Nit: extra newline.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java:
```diff
@@ -151,13 +154,15 @@
+        .caseSensitive(caseSensitive);
+
+    return nameMapping != null ?
+        builder.withNameMapping(NameMappingParser.fromJson(nameMapping)).build() : builder.build();
```
Same here, let's separate the name mapping from build.

chenjunjiedada replied:
Fixed.

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/MessageTypeToType.java:
```diff
@@ -229,7 +229,7 @@ private void addAlias(String name, int fieldId) {
   }
 
-  protected int nextId() {
+  private int nextId() {
```
I'd also prefer to revert this change. There may not be a use of it as a `protected` method any more, but we don't need to change this file and risk git conflicts.

chenjunjiedada replied:
OK

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/ApplyNameMapping.java:
```diff
@@ -0,0 +1,106 @@
+
+  @Override
+  public void beforeRepeatedElement(Type element) {
+  }
```
Can you add comments to these methods to explain why they are here?

chenjunjiedada replied:
OK, let me add the comments back.

rdblue commented:
@chenjunjiedada, looks like there are blockers with the new logic in `PruneColumns` and the test in `TestParquetSchemaUtil`. Otherwise, I think we can get this in. Please also fix the minor issues and style. Thanks!

rdblue commented:
Looks good! I would still like to fix the tests that don't remove IDs, but I'm going to merge this since it is a big commit to keep updating and to review. Thanks for all your work on this, @chenjunjiedada!
